Framework/PythonInterface/plugins/algorithms/CalculateEfficiencyCorrection.py

```diff
@@ -0,0 +1,256 @@
+
+        self.declareProperty(
+            name='Density',
+            defaultValue=0.0,
+            validator=FloatBoundedValidator(0.0),
+            doc='Mass density (g/cm^3) or Number density (atoms/Angstrom^3), :math:`\\rho`. \
+                 Default=0.0')
+
+        self.declareProperty(
+            name='Thickness',
+            defaultValue=1.0,
+            validator=FloatBoundedValidator(0.0),
+            doc='Sample thickness (cm), :math:`T`. Default value=1.0')
+
+        self.declareProperty(
+            name='MeasuredEfficiency',
+            defaultValue=0.0,
+            validator=FloatBoundedValidator(0.0, 1.0),
+            doc="Directly input the efficiency, :math:`\\epsilon`, measured at \
+                 MeasuredEfficiencyWavelength, :math:`\\lambda_{\\epsilon}`, to determine :math:`\\rho * T`, where \
+                 :math:`\\rho * T = - ln(1-\\epsilon) \\frac{1}{ \\frac{\\lambda_{\\epsilon} \\sigma (\\lambda_{ref})}{\\lambda_{ref}}}` \
+                 term for XSectionType == AttenuationXSection and \
+                 :math:`\\rho * T = - ln(1-\\epsilon) \\frac{1} \
+                    { \\sigma_s + \\frac{\\lambda_{\\epsilon} \\sigma (\\lambda_{ref})}{\\lambda_{ref}}}` \
+                 for XSectionType == TotalXSection \
+                 with :math:`\\lambda_{ref} =` 1.7982. Default=0.0")
+
+        self.declareProperty(
+            name='MeasuredEfficiencyWavelength',
+            defaultValue=1.7982,
+            validator=FloatBoundedValidator(0.0),
+            doc="The wavelength, :math:`\\lambda_{\\epsilon}`, at which the \
+                 MeasuredEfficiency, :math:`\\epsilon`, was measured. Default=1.7982")
+
+        self.declareProperty(
+            name='Alpha',
+            defaultValue=0.0,
+            doc="Directly input :math:`\\alpha`, where \
+                 :math:`\\alpha = \\rho * T * \\sigma (\\lambda_{ref}) / \\lambda_{ref}`, \
+                 where :math:`\\lambda_{ref} =` 1.7982. XSectionType has no effect. Default=0.0")
+
+        self.declareProperty(
+            name='XSectionType',
+            defaultValue="AttenuationXSection",
+            validator=StringListValidator(['AttenuationXSection', 'TotalXSection']),
+            doc = 'Use either the absorption cross section (for monitor-type correction) or, \
+                   the total cross section (for transmission-type correction) in Alpha term. \
+                   Default=AttenuationXSection')
+
+    def validateInputs(self):
+        issues = dict()
+
+        if self.getProperty('InputWorkspace').isDefault and self.getProperty('WavelengthRange').isDefault:
+            issues['InputWorkspace'] = "Must select either InputWorkspace and WavelengthRange as input"
+            issues['WavelengthRange'] = "Must select either InputWorkspace and WavelengthRange as input"
+
+        if not self.getProperty('InputWorkspace').isDefault and not self.getProperty('WavelengthRange').isDefault:
```

Comment:
I wonder if caching the default values would make the code a bit more readable:
```python
isInWSDefault = self.getProperty('InputWorkspace').isDefault
isWlRangeDefault = self.getProperty('WavelengthRange').isDefault
if isInWSDefault and isWlRangeDefault:
    pass
elif not isInWSDefault and not isWlRangeDefault:
   pass
```
and so on.

Reply:
I like this idea a lot, thanks!

Just a quick check, for the `not` comparisons do you think this is enough:
```
isDensityDefault = self.getProperty('Density').isDefault
isFormulaDefault = self.getProperty('ChemicalFormula').isDefault

if not isDensityDefault:
    if isFormulaDefault:
        issues['ChemicalFormula'] = "Must specify the ChemicalFormula with Density"
```

Or go as far as?:
```
isDensityDefault = self.getProperty('Density').isDefault
isFormulaDefault = self.getProperty('ChemicalFormula').isDefault
isDensitySetByUser = not self.getProperty('Density').isDefault

if isDensitySetByUser:
    if isFormulaDefault:
        issues['ChemicalFormula'] = "Must specify the ChemicalFormula with Density"
```

Reply:
I think the first option should be enough.